# Transcriptome profiling illuminates gene regulatory network underlying Brassica napus seed size development

Here we summarized the scripts used for the transcriptome profiling analyses conducted using two contrasting genotypes *B. napus cv.* Ragged Jack (RJ)  (small seed size) and B. napus var. Zhongshuang (ZS11)  (large seed size).

Please refer to the following zenodo link for proper citation:

[![DOI](https://github.com/Yedomon/bravo_project/blob/main/zenodo.4717617.svg)](https://zenodo.org/badge/latestdoi/xxxxx) 

The repository is organized as follows:

#### [01. QC, reads trimming, mapping and quantification](https://github.com/Yedomon/Yedomon-Genome_Assembly_Fusarium_oxysporum_f.sp._sesami/tree/main/01.Genome_size_estimation) - by Woolfenden, H. *et al.* A dataset of tissue-specific gene expression dynamics during seed development in Brassica. Figshare https://doi.org/10.6084/m9.figshare.27854361 (2024).

#### [02. DEG analysis]()

#### [03. Co-expression analysis]()


#### [04. Gene Regulatory Network (GRN) inference]()


#### [05. Seed size area boxplot]()


>  Scripts in this repository are licensed under General Public License v3.

>  For information inquiry please email Smita Kurup PhD. (smita.kurup@rothamsted.ac.uk)

> For reproducibility, please consider some changes in regard to the working directories.

